Fix for `create podidentityassociation` bug in auto-mode clusters

dheeraj-coding opened the pull request:
### Description

Added auto-mode checks for `IsPodIdentityAgentInstalled` to allow for creation of podidentityassociations in auto-mode clusters as pod-identity-agent comes in-built in auto-mode clusters and doesn't need explicit installs. 

Before:
`eksctl create podidentityassociation ...` in auto-mode cluster failed with the below errors
```
Error: the "eks-pod-identity-agent" addon must be installed to create pod identity associations; please enable it using `eksctl ....
```
but this is not expected behavior as auto-mode clusters come installed with pod-identity-agent and we should allow creation of pod-identity-associations. 

This PR:
Command successfully runs for auto-mode clusters and allows creation of pod-identity associations
```bash
2025-04-18 20:34:51 [ℹ]  1 task: {
    2 sequential sub-tasks: {
        create service account "default/eksctl-pr", if it does not already exist,
        create pod identity association for service account "default/eksctl-pr",
    } }2025-04-18 20:34:51 [ℹ]  created serviceaccount "default/eksctl-pr"
2025-04-18 20:34:51 [ℹ]  created pod identity association for service account "eksctl-pr" in namespace "default"
2025-04-18 20:34:51 [ℹ]  all tasks were completed successfully
```

Fixes #8346

Updated documentation to notify people that auto-mode clusters come pre-installed with pod-identity-agent

<img width="788" alt="Screenshot 2025-04-18 at 2 17 50 PM" src="https://github.com/user-attachments/assets/7ba02f18-f45d-4409-8121-ca586f9fd53a" />


### Checklist
- [ ] Added tests that cover your change (if possible)
- [&check;] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [&check;] Manually tested
- [&check;] Made sure the title of the PR is a good description that can go into the release notes
- [ ] (Core team) Added labels for change area (e.g. `area/nodegroup`) and kind (e.g. `kind/improvement`)

### BONUS POINTS checklist: complete for good vibes and maybe prizes?! :exploding_head:
- [ ] Backfilled missing tests for code in same general area :tada:
- [ ] Refactored something and made the world a better place :star2:

naclonts commented on userdocs/src/usage/pod-identity-associations.md:
```diff
@@ -10,6 +10,9 @@ As a result, IAM roles no longer need to reference an [OIDC provider](/usage/iam
 
 Behind the scenes, the implementation of pod identity associations is running an agent as a daemonset on the worker nodes. To run the pre-requisite agent on the cluster, EKS provides a new add-on called EKS Pod Identity Agent. Therefore, creating pod identity associations (in general, and with `eksctl`) requires the `eks-pod-identity-agent` addon pre-installed on the cluster. This addon can be [c ...
 
+???+ note
+    If you are using [EKS Auto Mode](https://eksctl.io/usage/auto-mode/) cluster the `eks-pod-identity-agent` comes pre-installed and you can skip creating the addon.
+
```
Looks like this may not work with Github's markdown format (screenshot of preview in Github PR). 

Also, should we put the note below the command on line 17? It's a little hard to see the connection between the paragraph and the CLI command with the note in between.

![image](https://github.com/user-attachments/assets/a45a6c0d-9507-496b-be4a-c9f71d066196)

naclonts commented on pkg/actions/podidentityassociation/migrator.go:
```diff
@@ -228,6 +233,22 @@ func IsPodIdentityAgentInstalled(ctx context.Context, eksAPI awsapi.EKS, cluster
 }
 
+func IsAutoModeEnabled(ctx context.Context, eksAPI awsapi.EKS, clusterName string) (bool, error) {
+	clstrDescribeResponse, err := eksAPI.DescribeCluster(ctx, &awseks.DescribeClusterInput{
+		Name: aws.String(clusterName),
+	})
+
+	if err != nil {
+		return false, fmt.Errorf("calling EKS::DescribeCluster: %w", err)
+	}
+
+	if *clstrDescribeResponse.Cluster.ComputeConfig.Enabled {
```
Nit:
```suggestion
	clustrDescribeResponse, err := eksAPI.DescribeCluster(ctx, &awseks.DescribeClusterInput{
		Name: aws.String(clusterName),
	})

	if err != nil {
		return false, fmt.Errorf("calling EKS::DescribeCluster: %w", err)
	}

	if *clustrDescribeResponse.Cluster.ComputeConfig.Enabled {
```

naclonts commented on pkg/actions/podidentityassociation/migrator_test.go:
```diff
@@ -67,6 +67,16 @@ var _ = Describe("Create", func() {
 	)
 
 	mockDescribeAddon := func(provider *mockprovider.MockProvider, err error) {
+		mockProvider.MockEKS().
```
Thanks for updating the test here! Do you think it makes sense to add a test case to ensure `IsPodIdentityAgentInstalled` returns true when auto-mode is enabled?

naclonts reviewed (commented):
Thanks for the PR @dheeraj-coding, this looks great! 

I left a few small suggestions/questions. It looks like the linter failed but just because go mod tidy needs to be ran:

```
go mod tidy && git commit go.mod go.sum --message "go mod tidy"
```

dheeraj-coding commented:
I have addressed most of the comments, but facing some build errors after merging from main and it is taking some time to fix. Hopefully I can come around soon to fix that.

dheeraj-coding commented:
I am not sure why unit-tests are failing locally for managed node groups. It is comparing some json files in ordered fashion but somehow the generated json files don't match the testdata. But I see current builds are all succeeding. I am not able to see the unit-tests are run or not in this PR. How can I check that?

dheeraj-coding commented:
ok looks like unit-tests in CI/CD do pass probably because they are caching. Can you take a look again? I haven't changed the notes section as I just replicated it from other notes section used already in the docs and they don't render either in md.

naclonts commented:
> ok looks like unit-tests in CI/CD do pass probably because they are caching. Can you take a look again? 

I think it looks all good, I notice the Karpenter tests fail locally for me (even on main) but work fine in CI, so I think that's a separate issue we need to fix for local development. 

Seems to be working well, so I'll go ahead and merge. Thanks for the contribution @dheeraj-coding!
